### PR TITLE
Biom t

### PR DESCRIPTION
**Pull request by ekopylova**
Script to convert Kraken output to BIOM Table.

### CODE REVIEW

File: python_scripts/parse_kraken_to_biom.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            if line.startswith('#'):
+                continue
+            line = line.strip().split("\t")
+            file_n = line[6].split(".bam")[0]
+            if file_n not in metadata:
+                metadata[file_n] = [line[7], line[5], line[9], line[0]]
+                num_samples += 1
+
+    organs_list = sorted({metadata[file_n][1] for file_n in metadata if isfile(join(taxonomy_report_dp, "%s_kraken.report_mpa" % file_n))})
+    print organs_list
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: is this repo going to be py3 compat?

File: python_scripts/parse_kraken_to_biom.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                      organs_list,
+                      reports_dp,
+                      taxa_level,
+                      taxonomy_report_dp,
+                      metadata):
+    """
+    """
+    output_biom_fp = join(reports_dp, "tcga_output_%s.biom" % (taxa_level))
+    with open(output_biom_fp, 'w') as output_f:
+        # output BIOM tsv file
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: :/ possible to just dump the abundances into `biom.Table`?

Comment (ElDeveloper) [inline python_scripts/parse_kraken_to_biom.py]: 👍

File: python_scripts/parse_kraken_to_biom.py (hunk 1 above)
Comment (ekopylova) [inline python_scripts/parse_kraken_to_biom.py]: of course!

File: python_scripts/parse_kraken_to_biom.py (hunk 2 above)
Comment (ekopylova) [inline python_scripts/parse_kraken_to_biom.py]: yes will do

**Comment by ekopylova** [comment]
@wasade could you take a look again :)?

@jkanbar this script will generate a BIOM table using results from Kraken translate command (as long as the sequences have QIIME labels).

File: python_scripts/parse_kraken_to_biom.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+    Parameters
+    ----------
+    table: biom.Table
+        an instance of a BIOM table
+    biom_output_fp: string
+        filepath to output BIOM table
+    """
+    with biom_open(biom_output_fp, 'w') as f:
+        if HAVE_H5PY:
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: can this check be omitted and just require the package to depend on h5py? BIOM-format 1.0 won't die if we keep using it...

File: python_scripts/parse_kraken_to_biom.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                       index):
+    """Absolute abundance of number of reads matching a defined taxa level.
+
+    Parameters
+    ----------
+    kraken_translate_report_fp: string
+        filepath to output of "kraken translate"
+    taxonomic_rank: string
+        taxonomy level (e.g., genus or species)
+    taxa_levels: dictionary
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: dictionary -> dict

File: python_scripts/parse_kraken_to_biom.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                       taxonomic_rank,
+                       taxa_levels,
+                       taxa_levels_idx,
+                       columns,
+                       index):
+    """Absolute abundance of number of reads matching a defined taxa level.
+
+    Parameters
+    ----------
+    kraken_translate_report_fp: string
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: string -> str

wont comment more on these, but the expectation is to use the type (although its not actually enforced anywhere that I know of)

File: python_scripts/parse_kraken_to_biom.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    taxonomic_rank_level_str = taxa_levels[taxonomic_rank]
+    taxonomic_rank_level_int = taxa_levels_idx[taxonomic_rank_level_str]
+    if taxonomic_rank_level_int < 6:
+        split_on_level = taxa_levels_idx[str(taxonomic_rank_level_int + 1)]
+    else:
+        # keep full string (to species level)
+        split_on_level = '\t'
+    with open(kraken_translate_report_fp) as kraken_translate_report_f:
+        for line in kraken_translate_report_f:
+            label, taxonomy = line.strip().split('\t')
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py:62]: would it be correct for this to be a `split('\t', 1)`?

File: python_scripts/parse_kraken_to_biom.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                taxonomy = taxonomy.split(split_on_level)[0]
+                sample_id = label.split('_')[0]
+                value = abundances.at[taxonomy, sample_id]
+                if np.isnan(value):
+                    abundances.set_value(taxonomy, sample_id, 1.)
+                else:
+                    abundances.set_value(taxonomy, sample_id, value+1.)
+    sample_ids = abundances.index.values.tolist()
+    for i in range(len(sample_ids)):
+        sample_ids[i] = sample_ids[i].replace("d__", "k__")
```
Comment (wasade) [inline python_scripts/parse_kraken_to_biom.py]: the sample IDs are taxonomy strings?

**Comment by wasade** [comment]
A few minor comments. This seems okay. The use of a `DataFrame` works, but it likely has more overhead than indexing into a `np.array` or `scipy.sparse.spmatrix`. However, I don't think that should block here

File: python_scripts/parse_kraken_to_biom.py (hunk 3 above)
Comment (ekopylova) [reply python_scripts/parse_kraken_to_biom.py]: yes

File: python_scripts/parse_kraken_to_biom.py (hunk 6 above)
Comment (ekopylova) [reply python_scripts/parse_kraken_to_biom.py:62]: hmm, it's a 2 column file with the 2nd column containing a string without spaces/tabs, so I don't think it's necessary?

File: python_scripts/parse_kraken_to_biom.py (hunk 7 above)
Comment (ekopylova) [reply python_scripts/parse_kraken_to_biom.py]: oops, no should be observation IDs

**Comment by ekopylova** [comment]
@wasade @jkanbar comments have been fixed, could you merge if everything looks ok?

**Pull request merged**